streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -0,0 +1,530 @@
+
+    private final SortedMap<TaskId, SortedSet<RankedClient<ID>>> statefulTasksToRankedCandidates;
+
+    public HighAvailabilityTaskAssignor(final Map<ID, ClientState> clientStates,
+                                 final Set<TaskId> allTasks,
+                                 final Set<TaskId> statefulTasks,
+                                 final AssignmentConfigs configs) {
+        this.configs = configs;
+        this.clientStates = clientStates;
+        this.allTasks = allTasks;
+        this.statefulTasks = new TreeSet<>(statefulTasks);
+
+        statelessTasks = new TreeSet<>(allTasks);
+        statelessTasks.removeAll(statefulTasks);
+
+        sortedClients = new TreeSet<>();
+        clientsToNumberOfThreads = new HashMap<>();
+        clientStates.forEach((client, state) -> {
+            sortedClients.add(client);
+            clientsToNumberOfThreads.put(client, state.capacity());
+        });
+
+        statefulTasksToRankedCandidates =
+            buildClientRankingsByTask(statefulTasks, clientStates, configs.acceptableRecoveryLag);
+    }
+
+    @Override
+    public boolean assign() {
+        final Map<ID, List<TaskId>> warmupTaskAssignment = initializeEmptyTaskAssignmentMap();
+        final Map<ID, List<TaskId>> standbyTaskAssignment = initializeEmptyTaskAssignmentMap();
+        final Map<ID, List<TaskId>> statelessActiveTaskAssignment = initializeEmptyTaskAssignmentMap();
+
+        // ---------------- Stateful Active Tasks ---------------- //
+
+        final Map<ID, List<TaskId>> statefulActiveTaskAssignment =
+            new DefaultStateConstrainedBalancedAssignor<ID>().assign(
+                statefulTasksToRankedCandidates,
+                configs.balanceFactor,
+                sortedClients,
+                clientsToNumberOfThreads
+            );
+
+        // ---------------- Warmup Replica Tasks ---------------- //
+
+        final Map<ID, List<TaskId>> balancedStatefulActiveTaskAssignment =
+            new DefaultBalancedAssignor<ID>().assign(
+                sortedClients,
+                statefulTasks,
+                clientsToNumberOfThreads,
+                configs.balanceFactor);
+
+        final Queue<Movement<ID>> movements = getMovements(statefulActiveTaskAssignment, balancedStatefulActiveTaskAssignment);
+        for (int numWarmupReplicas = 0; numWarmupReplicas < configs.maxWarmupReplicas; ++numWarmupReplicas) {
+            final Movement<ID> movement = movements.poll();
+            if (movement == null) {
+                break;
+            }
+            warmupTaskAssignment.get(movement.destination).add(movement.task);
+        }
+
+        // ---------------- Standby Replica Tasks ---------------- //
+
+        final List<Map<ID, List<TaskId>>> allStatefulTaskAssignments = asList(
+            statefulActiveTaskAssignment,
+            warmupTaskAssignment,
+            standbyTaskAssignment
+        );
+        final ClientValidatingPriorityQueue<ID> standbyTaskClientsQueue =
```

Comment:
Ultimately, I think we can and should refactor the standby task assignment out of `StickyTaskAssignor` and reuse the same strategy. But for now, I just put in this "good-enough" standby assignment approach so we can get the basic structure of the overall assignment algorithm down first.